kmk/bootcfg.py

```diff
@@ -0,0 +1,81 @@
+try:
+    from typing import Optional
+except ImportError:
+    pass
+
+import digitalio
+import microcontroller
+import usb_hid
+
+
+def bootcfg(
+    sense: [microcontroller.Pin, digitalio.DigitalInOut],
+    source: Optional[microcontroller.Pin, digitalio.DigitalInOut] = None,
```

Comment:
Calling it ```source``` is counter-intuitive, as diode would point away from it.

Reply:
I get what you mean. It's the "source" of the test voltage, and that name is commonly used as the opposite of a sense input.
Maybe I didn't explain it well enough in the documentation. I'm open to suggestions.

Reply:
Most people only interact with diodes by the means of their firmware config and thus are only aware of their abstract direction.

I think ```sense``` and ```source``` should be described separately then, like:
```sense``` the GPIO pin of directly wired key, or anode side of a key in matrix;
```source``` the GPIO pin of cathode side of a key in matrix.

Cathode and anode aren't terms commonly used in keyboard community, but they are easily googlabe.

Reply:
The whole diode point is something I intentionally avoided in the parameter description (let alone the technical terms of anode and cathode) and tried to explain only the basic eletrical meaning, leaving diode direction and direct wiring to the examples.
If you think that's something that should be mentioned in both places, or at least mentioning the examples in the parameter description, I'll have another go at it.

Reply:
The way I see it:
Requiring user to know what ```low``` is is bad;
Requiring used to reference an example is also bad, examples should be strictly supplementary IMO;
Dropping terms like anode and cathode is bad also, but a quick google search yields a picture with a neat arrow pointing away from anode and to cathode. Maybe include that diagram? ```anode ->|- cathode```

Reply:
Because `sense` and `source` are topology agnostic. The boot config pin isn't necessarily in a matrix, probably even a directly wired button on the MCU board.
I don't want to clutter the interface with lots of special cases that add complexity down the line.
Your pseudocode is not wrong, it's just incomplete and blows up in size if _all_ scenarios have to be handled.
This is as simple an interface as I can think of, and it's really not difficult to understand in my opinion -- if the documentation is improved.

Anyway, there's now more detailed documentation with configuration reference table.

Reply:
Alright, that's fair.